Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp

```diff
@@ -28,49 +28,369 @@
+{
+    return node.isTextNode() && node.renderer() && node.renderer()->style().visibility() == Visibility::Visible;
+}
+
+static std::optional<SimpleRange> findRangeFromNodeList(const String& query, SimpleRange& searchRange, Vector<RefPtr<Node>>& nodes, WordBounded wordStartBounded, WordBounded wordEndBounded)
+{
+    String searchBuffer;
+    std::optional<BoundaryPoint> start;
+    std::optional<BoundaryPoint> end;
+    
+    if (nodes.size() < 1)
         return std::nullopt;
+    
+    StringBuilder searchBufferBuilder;
+    for (auto node : nodes)
+        searchBufferBuilder.append(downcast<Text>(*node).data());
+    searchBuffer = searchBufferBuilder.toString();
+    
+    unsigned searchStart = 0;
+    
+    if (nodes[0] == &searchRange.startContainer())
+        searchStart = searchRange.startOffset();
+    
+    std::optional<unsigned> matchIndex;
+    
+    while (!matchIndex) {
+        // TODO: find only using the base characters i.e. also fold accents and others.
+        auto potentialIndex = searchBuffer.findIgnoringASCIICase(query, searchStart);
 
-    auto searchRange = makeRangeSelectingNodeContents(document);
+        if (potentialIndex == notFound)
+            return std::nullopt;
+        matchIndex = potentialIndex;
+        
+        start = getBoundaryPointAtIndex(*matchIndex, nodes, BoundaryPointIsAtEnd::No);
+        end = getBoundaryPointAtIndex(*matchIndex + query.length(), nodes, BoundaryPointIsAtEnd::Yes);
+        
+        if (((wordStartBounded == WordBounded::Yes) && !indexIsWordBoundary(searchBuffer, *matchIndex)) || ((wordEndBounded == WordBounded::Yes) && !indexIsWordBoundary(searchBuffer, *matchIndex + query.length()))) {
+            searchStart = *matchIndex + 1;
+            matchIndex = std::nullopt;
+        }
+    }
+        
+    unsigned endInset = 0;
+    if (nodes.last() == &searchRange.endContainer())
+        endInset = searchRange.endContainer().length() - searchRange.endOffset();
+    
+    if (*matchIndex + query.length() > searchBuffer.length() - endInset)
+        return std::nullopt;
+    
+    ASSERT_WITH_MESSAGE(start, "Scroll To Text Fragment: Start cannot be null");
+    ASSERT_WITH_MESSAGE(end, "Scroll To Text Fragment: End cannot be null");
     
-    FindOptions findOptions { DoNotRevealSelection };
+    ASSERT_WITH_MESSAGE(contains(searchRange, start.value()), "Scroll To Text Fragment: Start is within SearchRange.");
+    ASSERT_WITH_MESSAGE(contains(searchRange, end.value()), "Scroll To Text Fragment: End is within SearchRange.");
+    
+    if (!start || !end || !contains(searchRange, start.value()) || !contains(searchRange, end.value()))
+        return std::nullopt;
     
-    auto resultRange = collapseIfRootsDiffer(findPlainText(searchRange, target, findOptions));
+    return SimpleRange(*start, *end);
+    
+}
+
+static std::optional<SimpleRange> rangeOfStringInRange(const String& query, SimpleRange& searchRange, WordBounded wordStartBounded, WordBounded wordEndBounded)
```

Comment:
Can `searchRange` be `const` or is it intended to communicate out to the caller when it changes?

Reply:
We should name this function findStringInRange like in the spec text.
"range of string in range" sounds way too confusing.

Reply:
It most certainly need to be const here. Spec text (in general) passes in variables via copy.